lib-api/src/main/java/org/ergoplatform/appkit/babelfee/BabelFeeBoxState.java

```diff
@@ -0,0 +1,144 @@
+package org.ergoplatform.appkit.babelfee;
+
+import org.ergoplatform.appkit.ErgoId;
+import org.ergoplatform.appkit.ErgoToken;
+import org.ergoplatform.appkit.ErgoValue;
+import org.ergoplatform.appkit.InputBox;
+import org.ergoplatform.appkit.OutBox;
+import org.ergoplatform.appkit.OutBoxBuilder;
+import org.ergoplatform.appkit.Parameters;
+import org.ergoplatform.appkit.SigmaProp;
+import org.ergoplatform.appkit.TransactionBox;
+import org.ergoplatform.appkit.UnsignedTransactionBuilder;
+
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents a Babel Fee Box state, see EIP-0031
+ */
+public class BabelFeeBoxState {
+
+    private final long pricePerToken;
+    private final ErgoId tokenId;
+    private final SigmaProp boxCreator;
+    private final long value;
+    private final long tokenAmount;
+
+    public BabelFeeBoxState(TransactionBox ergoBox) {
+        value = ergoBox.getValue();
+        List<ErgoValue<?>> registers = ergoBox.getRegisters();
+        boxCreator = new SigmaProp((special.sigma.SigmaProp) registers.get(0).getValue());
+        pricePerToken = (Long) registers.get(1).getValue();
+        this.tokenId = new BabelFeeBoxContract().getTokenIdFromErgoTree(ergoBox.getErgoTree());
+
+        if (!ergoBox.getTokens().isEmpty()) {
+            ErgoToken ergoToken = ergoBox.getTokens().get(0);
+            tokenAmount = ergoToken.getValue();
+
+            if (!ergoToken.getId().equals(tokenId)) {
+                throw new IllegalStateException("token id of contract and token id in box diverge");
+            }
+        } else {
+            tokenAmount = 0;
+        }
+    }
+
+    BabelFeeBoxState(long pricePerToken, ErgoId tokenId, SigmaProp boxCreator, long value, long tokenAmount) {
+        this.pricePerToken = pricePerToken;
+        this.tokenId = tokenId;
+        this.boxCreator = boxCreator;
+        this.value = value;
+        this.tokenAmount = tokenAmount;
+    }
+
+    /**
+     * @return price offered per raw token amount
+     */
+    public long getPricePerToken() {
+        return pricePerToken;
+    }
+
+    /**
+     * @return token id this babel fee box is offering change for
+     */
+    public ErgoId getTokenId() {
+        return tokenId;
+    }
+
+    /**
+     * @return box creator or owner of the babel fee box
+     */
+    public SigmaProp getBoxCreator() {
+        return boxCreator;
+    }
+
+    /**
+     * @return overall ERG value in the box. not all is available to change for token as a small
+     * amount must remain in the box
```

Comment:
remain in which box? boxes are immutable.

Reply:
improved